### PR TITLE
feat: separate per-chain answerer task with configurable interval

### PR DESCRIPTION
**Pull request by luzzif**

### CODE REVIEW

File: src/answerer.rs [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+    defillama_http_client: Arc<HttpClient>,
+) -> anyhow::Result<()> {
+    let mut interval = interval(
+        chain_config
+            .answering_task_interval_seconds
+            .map(|seconds| Duration::from_secs(seconds))
+            .unwrap_or(ANSWERING_TASK_INTERVAL_SECONDS),
+    );
+
+    tracing::info!("answering active oracles");
```
Comment (guerrap) [inline src/answerer.rs]: Should we log the current `interval` being used here?

File: src/answerer.rs [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+    chain_id: u64,
+    signer: Arc<SignerMiddleware<Provider<Http>, LocalWallet>>,
+    db_connection_pool: Pool<ConnectionManager<PgConnection>>,
+    defillama_http_client: Arc<HttpClient>,
+) -> anyhow::Result<()> {
+    let mut db_connection = match db_connection_pool.get() {
+        Ok(connection) => connection,
+        Err(error) => {
+            tracing::error!("could not get new connection from pool: {:#}", error);
+            return Ok(());
```
Comment (guerrap) [inline src/answerer.rs:70]: Why `return Ok(())` and not just `Ok(())`? Same below.

File: src/answerer.rs [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+                    "could not get currently active oracles in chain with id {}: {:#}",
+                    chain_id,
+                    error
+                );
+                return Ok(());
+            }
+        };
+    drop(db_connection);
+
+    let active_oracles_len = active_oracles.len();
```
Comment (guerrap) [inline src/answerer.rs:88]: Should we just opt out here if `active_oracles_len == 0`?

File: src/answerer.rs [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+                    return Ok(());
+                }
+            }
+            answer
+        }
+    };
+    if let Some(answer) = answer {
+        // if we arrive here, an answer is available and we should submit it
+
+        tracing::info!("answering with value {}", answer);
```
Comment (guerrap) [inline src/answerer.rs:203]: Maybe we could add the oracle address in the logs here. and below too, no?

File: src/answerer.rs [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+        let debug_tx = format!("{:?}", tx);
+        let receipt = match tx.await {
+            Ok(receipt) => receipt,
+            Err(error) => {
+                // we need to throw the following errors as these needs to be addressed immediately.
+                // not being able to delete the tx hash for an oracle once an answer task errors out
+                // might cause a deadlock preventing any answering task from starting in the future
+
+                tracing::error!(
+                    "error while confirming answer transaction {}: {:#}",
```
Comment (guerrap) [inline src/answerer.rs:259]: Here too, we could add the oracle address in the error log.

File: src/answerer.rs (hunk 1 above)
Comment (luzzif) [reply src/answerer.rs]: Yes, sounds good. Addressed in dc626421fa88b6af86f2ea2d1b27b7bc7e4baa21

File: src/answerer.rs (hunk 2 above)
Comment (luzzif) [reply src/answerer.rs:70]: The second notation you provided only really works on the last instruction of the "main" execution path in a function/block. If you want to return early, you still need the return.

File: src/answerer.rs (hunk 3 above)
Comment (luzzif) [reply src/answerer.rs:88]: Addressed in df7929fc491aadbdce2a122c26a062dca3ac6bc5

File: src/answerer.rs (hunk 4 above)
Comment (luzzif) [reply src/answerer.rs:203]: In here we already have the oracle address in the logs in the spans section thanks to `instrument(info_span!("answer", chain_id, oracle_address))` on the answering task future.

File: src/answerer.rs (hunk 5 above)
Comment (luzzif) [reply src/answerer.rs:259]: Ditto

**Review by guerrap — APPROVED** [review]

**Pull request merged**